crates/starknet-devnet-core/src/starknet/forked_provider.rs

```diff
@@ -0,0 +1,108 @@
+use std::str::FromStr;
+use std::sync::{mpsc, Arc};
+
+use starknet_rs_core::types::BlockId;
+use starknet_rs_ff::FieldElement;
+use starknet_rs_providers::jsonrpc::HttpTransport;
+use starknet_rs_providers::{JsonRpcClient, Provider, ProviderError};
+use url::Url;
+
+use crate::error::{DevnetResult, Error, ForkedProviderError};
+
+type ProviderResult<T> = Result<T, ProviderError>;
+
+enum ForkedRequest {
+    GetNonceAt(FieldElement, std::sync::mpsc::Sender<ProviderResult<FieldElement>>),
+}
+
+struct ForkedProvider {
+    client: Arc<JsonRpcClient<HttpTransport>>,
+    receiver: tokio::sync::mpsc::Receiver<ForkedRequest>,
+}
+
+impl ForkedProvider {
+    fn new(fork_url: Url, rx: tokio::sync::mpsc::Receiver<ForkedRequest>) -> Self {
+        let json_rpc_client = JsonRpcClient::new(HttpTransport::new(fork_url));
+
+        Self { client: Arc::new(json_rpc_client), receiver: rx }
+    }
+
+    async fn run(mut self) {
+        while let Some(msg) = self.receiver.recv().await {
+            tokio::task::spawn(Self::handle_message(msg, self.client.clone()));
+        }
+    }
```

Comment:
What if there are multiple incoming get_nonce requests? Then they will all push into the same channel? Is that a problem when receiving from the channel?